pages/en/lb4/Testing-Your-application.md

```diff
@@ -8,676 +8,556 @@ permalink: /doc/en/lb4/Testing-your-application.html
-    ✓ morning prints good morning, bob
-    ✓ afternoon prints good afternoon, bob
-    ✓ evening prints good evening, bob
-    ✓ night prints good night, bob
-    ✓ greet prints hello, world
-    ✓ greet prints hello, bob
-
-
-  24 passing (87ms)
-```
+describe('Product (acceptance)', () => {
+  let app: HelloWorldApp;
+  let request: supertest.SuperTest<supertest.Test>;
 
-Takeaways:
+  before(givenEmptyDatabase);
+  before(givenRunningApp);
 
-1. Write the test first.
-1. Run the tests immediately and make sure it fails (to prevent false positives).
-1. Add integration tests to ensure compatibility between multiple units in your application.
-1. Integration tests do not guarantee your application still works as a whole (only the units under test), you still need acceptance tests!
-1. Easier to set up and get going than acceptance tests.
+  it('retrieves product details', async () => {
+    // arrange
+    const product = await givenProduct({
+      name: 'Ink Pen',
+      slug: 'ink-pen',
+      price: 1,
+      category: 'Stationery',
+      description: 'The ultimate ink-powered pen for daily writing',
+      label: 'popular',
+      available: true,
+      endDate: null,
+    });
 
-## Conclusion
+    // act
+    const response = await request.get('/product/ink-pen')
+
+    // assert
+    expect(response.body).to.deepEqual({
+      id: product.id,
+      name: 'Ink Pen',
+      slug: 'ink-pen',
+      price: 1,
+      category: 'Stationery',
+      available: true,
+      description: 'The ultimate ink-powered pen for daily writing',
+      label: 'popular',
+      endDate: null,
+    });
+  });
+
+  async function givenRunningApp() {
+    app = new HelloWorldApp();
+    const server = await app.getServer(RestServer);
+    server.bind(RestBindings.PORT).to(0);
+    await app.start();
+
+    const port: number = await server.get(RestBindings.PORT);
+    request = supertest(`http://127.0.0.1:${port}`);
+  }
+});
+```
 
-That's all there is to it. To summarize the entire process:
+### Test Sequence customizations
 
-1. Write a failing acceptance test.
-1. Write just enough code to make it pass.
-1. Write a failing unit test(s).
-1. Write just enough code to make it pass.
-1. Continue writing unit tests until satisfied.
-1. Write integration tests to ensure compatibility between multiple units.
-1. Write just enough code to make it pass.
-1. Continue writing integration tests until satisfied.
-1. Run the acceptance test to ensure everything continues to work together as a whole.
+Custom sequence behavior is best tested by observing changes in behavior of affected endpoints. For example, if your sequence involves authentication step that rejects anonymous requests for certain endpoints, then you can write a test making an anonymous request to such endpoint to verify that it's correctly rejected. These tests are esssentially the same as the test verifying implementation of individual endpoints as described in the previous section.
```

Comment:
I am not sure if this is the best advice, especially considering our own advice in [Thinking in LoopBack](https://github.com/strongloop/loopback.io/blob/0be8c43d389c62a0122eeb29cd04f338b2387aa1/pages/en/lb4/Thinking-in-LoopBack.md#how-to-build-a-great-test-suite):

> How to create a great test suite? By thinking smaller and favoring fast, focused unit-tests over slow application-wide end-to-end tests.

On one hand, writing several end-to-end tests to verify how auth is applied to each endpoint brings quite a lot of confidence for the security of our app. On the other hand, these tests are repetitive to write and take a long time to execute.

Perhaps we should look for better ways, for example:

 - A small number of integration tests to make sure all sequence customizations are applied for HTTP requests, these tests can invoke a dummy test controller method instead of real API endpoints - we are interested in the sequence, not any specific controller method.

 - Unit test for each controller method to ensure the method is correctly decorated with the metadata used by additional sequence actions used in our custom sequence.

For example, let's say we are using `@loopback/authentication` and its `@authenticate` decorator and `authenticate()` sequence action. To test such app, we would write:

 - four integration tests: two tests for a controller method decorated with `@authenticate` (an anonymous request, an authenticated request), two tests for a controller method not decorated with `@authenticate` (again, an anonymous request, an authenticated request)

 - for each controller method, one unit test verifying that the method is (or is not) decorated with `@authenticate`

@kjdelisle thoughts?

If we agree to rewrite this section, then I am proposing to do so in a follow-up pull request, so that this big patch can be landed ASAP.

Reply:
Opened a p2 issue to keep track of my suggestion: https://github.com/strongloop/loopback-next/issues/648